crates/energy/src/graph.rs

```diff
@@ -0,0 +1,239 @@
+const MAX_DISTANCE: f32 = 10.0;
+/// Minimum distance squared traveled by an object to update its nearby units.
+const MIN_DISTANCE_FOR_UPDATE: f32 = 0.5;
+
+pub(crate) struct GraphPlugin;
+
+impl Plugin for GraphPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugins(DespawnEventsPlugin::<&NearbyUnits, NearbyUnits>::default())
+            .add_systems(OnEnter(AppState::InGame), setup)
+            .add_systems(OnExit(AppState::InGame), clean_up)
+            .add_systems(
+                PostUpdate,
+                spawn_graph_components.run_if(in_state(AppState::InGame)),
+            )
+            .add_systems(
+                Update,
+                remove_old_nodes
+                    .before(GraphSystemSet::UpdateNearby)
+                    .run_if(in_state(GameState::Playing)),
+            )
+            .add_systems(
+                FixedUpdate,
+                (
+                    update_nearby.in_set(GraphSystemSet::UpdateNearby),
+                    update_graph
+                        .in_set(GraphSystemSet::UpdateGraph)
+                        .after(GraphSystemSet::UpdateNearby),
+                )
+                    .run_if(in_state(AppState::InGame)),
+            );
+
+        #[cfg(feature = "energy_graph_debug_lines")]
+        app.add_plugins(DebugLinesPlugin::default())
+            .add_systems(PostUpdate, debug_lines.run_if(in_state(AppState::InGame)));
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum GraphSystemSet {
+    UpdateNearby,
+    UpdateGraph,
+}
+
+/// wrapped entity to allow for default values (so we can work with TinyVec)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct NearbyEntity(Entity);
+
+impl Default for NearbyEntity {
+    fn default() -> Self {
+        Self(Entity::PLACEHOLDER)
+    }
+}
+
+impl From<NearbyEntity> for Entity {
+    fn from(val: NearbyEntity) -> Self {
+        val.0
+    }
+}
+
+impl From<Entity> for NearbyEntity {
+    fn from(entity: Entity) -> Self {
+        Self(entity)
+    }
+}
+
+/// The power grid resource is used to store the power grid graph.
+#[derive(Resource, Debug, Clone)]
+pub(crate) struct PowerGrid {
+    /// The power grid graph.
+    graph: GraphMap<Entity, f64, Undirected>,
+}
+
+impl Default for PowerGrid {
+    fn default() -> Self {
+        Self {
+            graph: GraphMap::new(),
+        }
+    }
+}
+
+/// The energy grid member component is used to store the energy grid member entities.
+#[derive(Component, Debug, Clone)]
+pub struct EnergyGridMember;
+
+/// The nearby units component is used to store the nearby entities of an entity.
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits {
+    units: TinyVec<[NearbyEntity; 16]>,
+    last_pos: Option<Vec2>,
+}
+
+impl NearbyUnits {
+    fn remove_matching(&mut self, entity: NearbyEntity) {
+        self.units.retain(|e| *e != entity);
```

Comment:
Finding index of the entity and using swap-remove might be faster. Retain has to shift all following entities and thus is somewhat slower.